### PR TITLE
[Merged by Bors] - feat: allow differential objects with shifts in an AddMonoidWithOne

### PR DESCRIPTION
**Pull request by kim-em**
What I was really hoping to do was show that ℕ-graded DGOs were the same as ℕ-graded chain complexes (we have the ℤ version), but this still seems awkward. If anyone would like to help with that, please let me know. I think this might be helpful in constructor tensor products of chain complexes with less suffering.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Homology/DifferentialObject.lean [hunk 1]
```diff
@@ -146,7 +146,7 @@ to the category of homological complexes in `V`.
 -/
 @[simps]
 def dgoEquivHomologicalComplex :
-    DifferentialObject (GradedObjectWithShift b V) ≌ HomologicalComplex V (ComplexShape.up' b) where
+    DifferentialObject ℤ (GradedObjectWithShift b V) ≌ HomologicalComplex V (ComplexShape.up' b) where
```
Comment (semorrison) [inline Mathlib/Algebra/Homology/DifferentialObject.lean:149]:
```suggestion
    DifferentialObject ℤ (GradedObjectWithShift b V) ≌ 
      HomologicalComplex V (ComplexShape.up' b) where
```

File: Mathlib/CategoryTheory/DifferentialObject.lean [hunk 2]
```diff
@@ -68,94 +68,96 @@ namespace Hom
...
   Hom.ext _ _ w
 
 @[simp]
-theorem id_f (X : DifferentialObject C) : (𝟙 X : X ⟶ X).f = 𝟙 X.obj := rfl
+theorem id_f (X : DifferentialObject S C) : (𝟙 X : X ⟶ X).f = 𝟙 X.obj := rfl
 #align category_theory.differential_object.id_f CategoryTheory.DifferentialObject.id_f
 
 @[simp]
-theorem comp_f {X Y Z : DifferentialObject C} (f : X ⟶ Y) (g : Y ⟶ Z) : (f ≫ g).f = f.f ≫ g.f := rfl
+theorem comp_f {X Y Z : DifferentialObject S C} (f : X ⟶ Y) (g : Y ⟶ Z) : (f ≫ g).f = f.f ≫ g.f := rfl
```
Comment (semorrison) [inline Mathlib/CategoryTheory/DifferentialObject.lean:99]:
```suggestion
theorem comp_f {X Y Z : DifferentialObject S C} (f : X ⟶ Y) (g : Y ⟶ Z) : 
    (f ≫ g).f = f.f ≫ g.f := 
  rfl
```

File: Mathlib/CategoryTheory/DifferentialObject.lean [hunk 3]
```diff
@@ -28,10 +28,10 @@ universe v u
 
 namespace CategoryTheory
 
-variable (C : Type u) [Category.{v} C]
+variable (S : Type) [AddMonoidWithOne S] (C : Type u) [Category.{v} C]
 
--- TODO: generalize to `HasShift C A` for an arbitrary `[AddMonoid A]` `[One A]`.
-variable [HasZeroMorphisms C] [HasShift C ℤ]
+-- TODO: generalize to `HasShift C S` for an arbitrary `[AddMonoidWithOne S]`
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/DifferentialObject.lean:33]: Can this TODO be removed now?

File: Mathlib/CategoryTheory/DifferentialObject.lean [hunk 4]
```diff
@@ -28,10 +28,10 @@ universe v u
 
 namespace CategoryTheory
 
-variable (C : Type u) [Category.{v} C]
+variable (S : Type) [AddMonoidWithOne S] (C : Type u) [Category.{v} C]
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/DifferentialObject.lean:31]: Did you intentionally restrict the universe of `S` here and elsewhere in this file? The file still seems to compile after generalizing.

File: Mathlib/CategoryTheory/DifferentialObject.lean [hunk 5]
```diff
@@ -174,19 +178,19 @@ universe v' u'
 
 variable (D : Type u') [Category.{v'} D]
 
-variable [HasZeroMorphisms D] [HasShift D ℤ]
+variable [HasZeroMorphisms D] [HasShift D S]
 
 /-- A functor `F : C ⥤ D` which commutes with shift functors on `C` and `D` and preserves zero
 morphisms can be lifted to a functor `DifferentialObject C ⥤ DifferentialObject D`. -/
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/DifferentialObject.lean:184]:
```suggestion
morphisms can be lifted to a functor `DifferentialObject S C ⥤ DifferentialObject S D`. -/
```

File: Mathlib/CategoryTheory/DifferentialObject.lean [hunk 6]
```diff
@@ -245,15 +251,15 @@ end DifferentialObject
...
 
-variable (C : Type u) [Category.{v} C]
+variable {S : Type} [AddCommGroupWithOne S] (C : Type u) [Category.{v} C]
 
-variable [HasZeroMorphisms C] [HasShift C ℤ]
+variable [HasZeroMorphisms C] [HasShift C S]
 
 noncomputable section
 
 /-- The shift functor on `DifferentialObject C`. -/
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/DifferentialObject.lean:260]:
```suggestion
/-- The shift functor on `DifferentialObject S C`. -/
```

File: Mathlib/CategoryTheory/DifferentialObject.lean (hunk 3 above)
Comment (semorrison) [inline Mathlib/CategoryTheory/DifferentialObject.lean:33]:
```suggestion
```

File: Mathlib/CategoryTheory/DifferentialObject.lean (hunk 4 above)
Comment (semorrison) [inline Mathlib/CategoryTheory/DifferentialObject.lean:31]:
```suggestion
variable (S : Type _) [AddMonoidWithOne S] (C : Type u) [Category.{v} C]
```

Comment (semorrison) [reply Mathlib/CategoryTheory/DifferentialObject.lean:31]: I'd hope that we don't ever have to use this for something other that Nat or Int, but you're right, we may as well. :-)

File: Mathlib/CategoryTheory/DifferentialObject.lean [hunk 7]
```diff
@@ -210,13 +213,14 @@ namespace CategoryTheory
 
 namespace DifferentialObject
 
-variable (C : Type u) [Category.{v} C]
+variable (S : Type) [AddMonoidWithOne S] (C : Type u) [Category.{v} C]
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/DifferentialObject.lean:216]:
```suggestion
variable (S : Type _) [AddMonoidWithOne S] (C : Type u) [Category.{v} C]
```

File: Mathlib/CategoryTheory/DifferentialObject.lean [hunk 8]
```diff
@@ -245,15 +250,15 @@ end DifferentialObject
 
 namespace DifferentialObject
 
-variable (C : Type u) [Category.{v} C]
+variable {S : Type} [AddCommGroupWithOne S] (C : Type u) [Category.{v} C]
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/DifferentialObject.lean:253]:
```suggestion
variable {S : Type _} [AddCommGroupWithOne S] (C : Type u) [Category.{v} C]
```

File: Mathlib/CategoryTheory/DifferentialObject.lean [hunk 9]
```diff
@@ -228,15 +232,16 @@ end DifferentialObject
 
 namespace DifferentialObject
 
+variable (S : Type) [AddMonoidWithOne S]
```
Comment (TwoFX) [inline Mathlib/CategoryTheory/DifferentialObject.lean:235]:
```suggestion
variable (S : Type _) [AddMonoidWithOne S]
```

**Review by TwoFX — APPROVED** [review]

**Comment by TwoFX** [comment]
bors d+

**Comment by semorrison** [comment]
bors merge

**Pull request closed without merging**